### DISCUSSION TITLE
Override individual template defaults from the command line

### DISCUSSION THREAD

**Discussion by op-ct** [puppetlabs-toy-chest/pct#134] [Ideas]
I'd rather supply one-off template default overrides from the command line than stop to make an override file each time.

Would it be possible to just tack on an extra options as k/v pairs, like how Bolt handles parameters?

Using the  `example_template` from https://github.com/puppetlabs/pdkgo#overriding-template-defaults as an example:

```sh
pct new example_template --name foo example_template.isPuppet=false
```

**Comment by jpogran** [comment]
Thanks for the feedback! I really appreciate asking about this, as this is something we've been working extra hard to get right. 
Do you have an example of the type of template that would have config values that would need a value like this replaced without wanting that customization for any other run of that template?

The challenge here is that we allow a schema-less k/v set of data to apply to the templates. This is great because it allows a huge set of flexibility in how to model the template to deploy and it's relatively easy to write down in yaml, all without having to have our team approve a PR.

It's not-so-great if you want to model it at the command line. Your example works well, but how about a nested array or similar complicated structure? There is also all the different shells that we support, which may make that syntax hard or impossible to write down with their different parsing routines. Shell escaping characters is an art form, one that we want to try to avoid. I know bolt supports some k/v pairs on the command line for parts of it's app, and I also know the places where it doesn't work out so well at the commandline. I'm trying to avoid these problems with this project.

That being said, we *are* trying out different ideas. We added https://github.com/puppetlabs/pdkgo/pull/115 as a lighter weight approach to context specific data. It's not directly related to your question here, but is another option. Another approach we initially thought to try, but de-prioritized for now, is allowing prefixed environment variables. Something like: `PCT_EXAMPLE_TEMPLATE.ISPUPPET=false pct new example_template --name foo`. The configuration library we're using requires the environment variables to be prefixed and upper case, but also allows us to write custom transformations to pull it back to a format we expect.

**Comment by da-ar** [comment]
https://github.com/puppetlabs/pdkgo/pull/115 is merged and workspace configuration provides this functionality.
